### PR TITLE
Add EDF IO

### PR DESCRIPTION
**Pull request by JuliaSprenger**
This IO supports reading files in the European Data Format (EDF, https://www.edfplus.info/). For now only original EDF files and EDF+ files with continuous data (EDF+C) are accepted. The IO is based on the pyedflib package.

### CODE REVIEW

**Comment by samuelgarcia** [comment]
Hi Julia.
This is super cool.
From what I see, this is not lazy!!!
The entire signals is loaded in memory no ?

**Comment by JuliaSprenger** [comment]
Hi Sam, yes I think the pyedflib highlevel support does not allow for memory mapping. For now this at least allows loading of EDF files in the first place. Performance optimization can be added in a 2nd step.

**Comment by JuliaSprenger** [comment]
I updated the PR to handle multiple streams, data loading only on request and usage of int16 for data handling as in the edf file format.

**Comment by JuliaSprenger** [comment]
Hi @samuelgarcia I added a __del__ method, but think that the EDFIO tests  are still failing as the garbage collection is not run between each of the tests methods. Any idea on how to solve this?

**Comment by sportnoah14** [comment]
Is there currently something that needs assistance on? I'm very much looking forward to edf support. Just as a not, I've worked with pyedflib in the past and it can be very finicky. Especially when there's gaps in time. This repo is a pretty good reference/resource as it strips reading and writing edf files into manageable and easy to read python code. https://github.com/bids-standard/pyedf

**Comment by JuliaSprenger** [comment]
Hi @sportnoah14! Thanks for reaching out. For now I have limited the IO to read only continuous data files (edf+C or edf), so there should be no gaps (or did you mean another type of gap here?) Would this be sufficient also for your case?

If you have some real data edf files (in contrast to the simulated ones I am working with right now) would you be interested in providing an tiny file for testing purposes?

**Comment by sportnoah14** [comment]
@JuliaSprenger , I'd be happy to send example data. Sometimes edf files can contain gaps in time. Meaning that there may be a few seconds or more of missing data. For this, from my experience, pyedflib throws an error. Do you have a dropbox or google drive I can upload example data to? I'll be sure to include some data with gaps in time and/or that I've noticed doesn't play nice with pyedflib

**Comment by JuliaSprenger** [comment]
Heyho! You can deposit you files [here](https://amubox.univ-amu.fr/s/5Cd2BedrfJs9BQf) and I will try to update the EDFIO to cover all the cases

**Comment by JuliaSprenger** [comment]
Hi @sportnoah14 If you don't want to upload your files with the provided link you can also create small test files (<10MB) and open a direct PR towards our collection of test data files (https://gin.g-node.org/NeuralEnsemble/ephy_testing_data), see detailed contribution instructions in the [Readme](https://gin.g-node.org/NeuralEnsemble/ephy_testing_data/src/master/README.md)

**Comment by JuliaSprenger** [comment]
I resolved the file closing issue in the tests and added support for edf annotations (neo events and epochs). 
@sportnoah14 Thanks for providing the file. I used it to add the support for the edf annotations. However this file is unfortunately still too big to be used as a public test file. Do you also have a smaller version of the files that still contains edf annotations, ideally <<10MB.

**Comment by JuliaSprenger** [comment]
The latest updates build on top of the improved test files (https://gin.g-node.org/NeuralEnsemble/ephy_testing_data/pulls/76)

**Comment by JuliaSprenger** [comment]
@samuelgarcia I think this is a first good version that could be merged. With more test file becoming available hopefully later I can cover more improvements and new test file integration in a separate PR. What do you think?

**Comment by JuliaSprenger** [comment]
@samuelgarcia What do you think?

File: neo/io/edfio.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+"""
+
+from neo.io.basefromrawio import BaseFromRaw
+from neo.rawio.edfrawio import EDFRawIO
+
+
+class EDFIO(EDFRawIO, BaseFromRaw):
+    """
+    IO for reading edf and edf+ files.
+    Note that this IO does not yet implement true lazy loading.
```
Comment (JuliaSprenger) [inline neo/io/edfio.py:20]: Remove misleading documentation
```suggestion
```

**Comment by JuliaSprenger** [comment]
@samuelgarcia ping?

**Comment by JuliaSprenger** [comment]
@samuelgarcia ping?

File: neo/rawio/edfrawio.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        for i, channel_idx in enumerate(selected_channel_idxs):
+            # use int32 for compatibility with pyedflib
+            buffer = np.empty(n, dtype=np.int32)
+            self.edf_reader.read_digital_signal(channel_idx, i_start, n, buffer)
+            data.append(buffer)
+
+        # downgrade to int16 as this is what is used in the edf file format
+        data = np.asarray(data, dtype=np.int16)
+
+        return data.T  # use dimensions (time, channel)
```
Comment (JuliaSprenger) [inline neo/rawio/edfrawio.py:203]: transposing should be happening before casting

File: neo/rawio/edfrawio.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+        t_stop = self.edf_reader.datarecord_duration * self.edf_reader.datarecords_in_file
+        # this must return an float scale in second
+        return t_stop
+
+    def _get_signal_size(self, block_index, seg_index, stream_index):
+        chidx = self.stream_idx_to_chidx[stream_index][0]
+        # use sample count of first signal in stream
+        return self.edf_reader.getNSamples()[chidx]
+
+    def _get_signal_t_start(self, block_index, seg_index, stream_index):
```
Comment (JuliaSprenger) [inline neo/rawio/edfrawio.py:168]: this should return float instead of integer

**Comment by JuliaSprenger** [comment]
Hi @samuelgarcia I implemented the changes we discussed. Could you review the rest?

**Review by samuelgarcia — APPROVED** [review]

**Comment by betty6you** [comment]
Hello,
I am using this to extract the info from EDF file but it keeps giving me error saying '**file has already been opened**'

Is there a close function missing in the code?

Hope you can help me solve it. 

Attached is the code. https://github.com/betty6you/EDFRecordingExtractor/blob/main/EDF_extractor_test.ipynb

Many thanks!

**Comment by samuelgarcia** [comment]
Hi,
thanks for this.
Oups that is a problem.
The edf lib used interanally by neo forgive to open the open the file several time but spikeinterface that used neo by design open the file several time.

The only solution I see is to store your file in a new format from the beginig in spikeinetrface and use that format for more processing.

```python
recording=se.EDFRecordingExtractor(recording_file, stream_id=None, stream_name=None, all_annotations=False)
recording2 = recording.save(folder='/path/to/new/folder')
```

**Comment by betty6you** [comment]
Thanks for your reply!
I will give a try on the save method.

So for the later processing, I need to use recording2 from now on?

Thanks!

Sent from my iPhone

On Oct 10, 2022, at 3:27 AM, Garcia Samuel ***@***.***> wrote:

﻿

Hi,
thanks for this.
Oups that is a problem.
The edf lib used interanally by neo forgive to open the open the file several time but spikeinterface that used neo by design open the file several time.

The only solution I see is to store your file in a new format from the beginig in spikeinetrface and use that format for more processing.

recording=se.EDFRecordingExtractor(recording_file, stream_id=None, stream_name=None, all_annotations=False)
recording2 = recording.save(folder='/path/to/new/folder')

**Comment by samuelgarcia** [comment]
yes exactly recording2 is the same but in the format binary. Which the more easy to handle.

**Comment by betty6you** [comment]
Thanks. I tried this method, it worked well. 
But what if I need to run multiple files at a time? The save method will still work?

**Pull request merged**